### PR TITLE
SUV conversion for the GML units 

### PR DESCRIPTION
**Pull request by maksymfritsak**
According to the SUV manuscript

### CODE REVIEW

**Comment by hubertgabrys** [comment]
How about such a revision to the calculate_suv()? I haven't tested this code, but I think such code organization is a bit more readible. What do you think? I also added errors for configurations that are not supported.

```python
    def calculate_suv(dicom_file_path, min_acquisition_time):
        ds = pydicom.dcmread(dicom_file_path)
        units = ds.Units
        image_data = ds.pixel_array
        activity_concentration = (image_data * ds.RescaleSlope) + ds.RescaleIntercept

        if units == 'GML':
            suv = activity_concentration

        elif units == 'BQML':
            injection_time = parse_time(ds.RadiopharmaceuticalInformationSequence[0].RadiopharmaceuticalStartTime)
            patient_weight = float(ds.PatientWeight)
            injected_dose = float(ds.RadiopharmaceuticalInformationSequence[0].RadionuclideTotalDose)
            half_life = float(ds.RadiopharmaceuticalInformationSequence[0].RadionuclideHalfLife)

            if ds.DecayCorrection == 'START':
                if 'PHILIPS' in ds.Manufacturer.upper():
                    acquisition_time = min_acquisition_time
                elif 'SIEMENS' in ds.Manufacturer.upper():
                    try:
                        acquisition_time = parse_time(ds[(0x0071, 0x1022)].value).replace(year=injection_time.year,
                                                                                          month=injection_time.month,
                                                                                          day=injection_time.day)
                    except KeyError:
                        acquisition_time = min_acquisition_time
                elif 'GE' in ds.Manufacturer.upper():
                    try:
                        acquisition_time = parse_time(ds[(0x0009, 0x100d)].value).replace(year=injection_time.year,
                                                                                          month=injection_time.month,
                                                                                          day=injection_time.day)
                    except KeyError:
                        acquisition_time = min_acquisition_time
                else:
                    error_msg = f"Vendor {ds.Manufacturer} is not supported with units {units}!"
                    raise DataStructureError(error_msg)
            elif ds.DecayCorrection == 'ADMIN':
                acquisition_time = injection_time
            else:
                error_msg = f"Decay correction {ds.DecayCorrection} is not supported!"
                raise DataStructureError(error_msg)
            elapsed_time = (acquisition_time - injection_time).total_seconds()
            decay_factor = np.exp(-1 * ((np.log(2) * elapsed_time) / half_life))
            decay_corrected_dose = injected_dose * decay_factor
            suv = activity_concentration / (decay_corrected_dose / (patient_weight * 1000))

        elif units == 'CNTS':
            if 'PHILIPS' in ds.Manufacturer.upper():
                suv = activity_concentration * ds[(0x7053, 0x1009)].value
            else:
                error_msg = f"Vendor {ds.Manufacturer} is not supported with units {units}!"
                raise DataStructureError(error_msg)

        else:
            error_msg = f"Units {units} are not supported!"
            raise DataStructureError(error_msg)

        return suv.T
```

**Comment by maksymfritsak** [comment]
I need to check this, but it looks like the Philips strategy is wrong now. As activity_concentration * ds[(0x7053, 0x1009)].value should give you a corrected activity concentration and not an SUV.

**Comment by maksymfritsak** [comment]
It should be like this then: 

elif units == 'CNTS':
            if 'PHILIPS' in ds.Manufacturer.upper():
                acquisition_time = min_acquisition_time
                elapsed_time = (acquisition_time - injection_time).total_seconds()
               decay_factor = np.exp(-1 * ((np.log(2) * elapsed_time) / half_life))
               decay_corrected_dose = injected_dose * decay_factor
                suv = (activity_concentration * ds[(0x7053, 0x1009)].value) / (decay_corrected_dose / (patient_weight * 1000))
            else:
                error_msg = f"Vendor {ds.Manufacturer} is not supported with units {units}!"
                raise DataStructureError(error_msg)

**Comment by maksymfritsak** [comment]
This should work, I hope :)

`    def calculate_suv(dicom_file_path, min_acquisition_time):
        ds = pydicom.dcmread(dicom_file_path)
        units = ds.Units
        image_data = ds.pixel_array
        activity_concentration = (image_data * ds.RescaleSlope) + ds.RescaleIntercept

        if units == 'GML':
            suv = activity_concentration

        elif units == 'BQML':
            injection_time = parse_time(ds.RadiopharmaceuticalInformationSequence[0].RadiopharmaceuticalStartTime)
            patient_weight = float(ds.PatientWeight)
            injected_dose = float(ds.RadiopharmaceuticalInformationSequence[0].RadionuclideTotalDose)
            half_life = float(ds.RadiopharmaceuticalInformationSequence[0].RadionuclideHalfLife)

            if ds.DecayCorrection == 'START':
                if 'PHILIPS' in ds.Manufacturer.upper():
                    acquisition_time = min_acquisition_time
                elif 'SIEMENS' in ds.Manufacturer.upper():
                    try:
                        acquisition_time = parse_time(ds[(0x0071, 0x1022)].value).replace(year=injection_time.year,
                                                                                          month=injection_time.month,
                                                                                          day=injection_time.day)
                    except KeyError:
                        acquisition_time = min_acquisition_time
                elif 'GE' in ds.Manufacturer.upper():
                    try:
                        acquisition_time = parse_time(ds[(0x0009, 0x100d)].value).replace(year=injection_time.year,
                                                                                          month=injection_time.month,
                                                                                          day=injection_time.day)
                    except KeyError:
                        acquisition_time = min_acquisition_time
                else:
                    error_msg = f"Vendor {ds.Manufacturer} is not supported with units {units}!"
                    raise DataStructureError(error_msg)
            elif ds.DecayCorrection == 'ADMIN':
                acquisition_time = injection_time
            else:
                error_msg = f"Decay correction {ds.DecayCorrection} is not supported!"
                raise DataStructureError(error_msg)
            elapsed_time = (acquisition_time - injection_time).total_seconds()
            decay_factor = np.exp(-1 * ((np.log(2) * elapsed_time) / half_life))
            decay_corrected_dose = injected_dose * decay_factor
            suv = activity_concentration / (decay_corrected_dose / (patient_weight * 1000))

        elif units == 'CNTS':
            if 'PHILIPS' in ds.Manufacturer.upper():
                acquisition_time = min_acquisition_time
                elapsed_time = (acquisition_time - injection_time).total_seconds()
                decay_factor = np.exp(-1 * ((np.log(2) * elapsed_time) / half_life))
                decay_corrected_dose = injected_dose * decay_factor
                suv = (activity_concentration * ds[(0x7053, 0x1009)].value) / (decay_corrected_dose / (patient_weight * 1000))
            else:
                error_msg = f"Vendor {ds.Manufacturer} is not supported with units {units}!"
                raise DataStructureError(error_msg)

        else:
            error_msg = f"Units {units} are not supported!"
            raise DataStructureError(error_msg)

        return suv.T`

**Comment by hubertgabrys** [comment]
Ah right. It's because we use f_bqml not f_suv for Philips.

**Comment by hubertgabrys** [comment]
I think there's still a bit of duplicated code and we could improve the modularity and readibility. I also suggest to rename function 'calculate_suv' to 'process_single_slice' or 'calculate_suv_in_slice' to make it clear that it does.

What do you think about sth like this? I haven't tested it, so it's more about giving you an idea of what I had in mind.

```python
    def process_gml(pixel_array_units):
        return pixel_array_units

    def process_bqml(activity_concentration, ds, min_acquisition_time):
        injection_time = parse_time(ds.RadiopharmaceuticalInformationSequence[0].RadiopharmaceuticalStartTime)
        patient_weight = float(ds.PatientWeight)
        injected_dose = float(ds.RadiopharmaceuticalInformationSequence[0].RadionuclideTotalDose)
        half_life = float(ds.RadiopharmaceuticalInformationSequence[0].RadionuclideHalfLife)
        manufacturer = ds.Manufacturer.upper()

        if ds.DecayCorrection == 'START':
            if 'PHILIPS' in manufacturer:
                acquisition_time = min_acquisition_time
            elif 'SIEMENS' in manufacturer:
                try:
                    acquisition_time = parse_time(ds[(0x0071, 0x1022)].value).replace(year=injection_time.year,
                                                                                      month=injection_time.month,
                                                                                      day=injection_time.day)
                except KeyError:
                    acquisition_time = min_acquisition_time
            elif 'GE' in manufacturer:
                try:
                    acquisition_time = parse_time(ds[(0x0009, 0x100d)].value).replace(year=injection_time.year,
                                                                                      month=injection_time.month,
                                                                                      day=injection_time.day)
                except KeyError:
                    acquisition_time = min_acquisition_time
            else:
                error_msg = f"Vendor {ds.Manufacturer} is not supported with BQML units!"
                raise DataStructureError(error_msg)
        elif ds.DecayCorrection == 'ADMIN':
            acquisition_time = injection_time
        else:
            error_msg = f"Decay correction {ds.DecayCorrection} is not supported!"
            raise DataStructureError(error_msg)

        elapsed_time = (acquisition_time - injection_time).total_seconds()
        decay_factor = np.exp(-1 * ((np.log(2) * elapsed_time) / half_life))
        decay_corrected_dose = injected_dose * decay_factor
        suv = activity_concentration / (decay_corrected_dose / (patient_weight * 1000))

        return suv

    def process_cnts(pixel_array_units, ds, min_acquisition_time):
        if 'PHILIPS' in ds.Manufacturer.upper():
            activity_concentration_bqml = pixel_array_units * ds[(0x7053, 0x1009)].value
            suv = process_bqml(activity_concentration_bqml, ds, min_acquisition_time)
        else:
            error_msg = f"Vendor {ds.Manufacturer} is not supported with CNTS units!"
            raise DataStructureError(error_msg)
        return suv

    def process_single_slice(dicom_file_path, min_acquisition_time):
        ds = pydicom.dcmread(dicom_file_path)
        units = ds.Units
        pixel_array_units = (ds.pixel_array * ds.RescaleSlope) + ds.RescaleIntercept

        if units == 'GML':
            suv = process_gml(pixel_array_units)
        elif units == 'BQML':
            suv = process_bqml(pixel_array_units, dicom_file, min_acquisition_time)
        elif units == 'CNTS':
            suv = process_cnts(pixel_array_units, dicom_file, min_acquisition_time)
        else:
            error_msg = f"Units {units} are not supported!"
            raise DataStructureError(error_msg)

        return suv.T
 ```

**Comment by maksymfritsak** [comment]
This should work!

File: zrad/logic/image.py [hunk 1]
```diff
@@ -262,66 +262,98 @@ def validate_pet_dicom_tags(dicom_files):
...
-            activity_concentration = activity_concentration * ds[(0x7053, 0x1009)].value
+        ds = pydicom.dcmread(dicom_file_path)
+        units = ds.Units
+        pixel_array_units = (ds.pixel_array * ds.RescaleSlope) + ds.RescaleIntercept
 
-        suv = activity_concentration / (decay_corrected_dose / (patient_weight * 1000))
+        if units == 'GML':
+            if (0x0054, 0x1006) in ds:
+                if ds[0x0054, 0x1006].value == 'BW':
+                    suv = process_gml(pixel_array_units)
```
Comment (hubertgabrys) [inline zrad/logic/image.py:344]:
```suggestion
                if ds[0x0054, 0x1006].value == 'BW':
                    suv = process_gml(pixel_array_units)
                else:
                    error_msg = f"GML with {ds[0x0054, 0x1006].value} SUV normalizatoin is not supported!"
                    raise DataStructureError(error_msg)
```

**Review by hubertgabrys — APPROVED** [review]

**Pull request merged**